hal/armv7m/imxrt/117x/imxrt.c

```diff
@@ -389,6 +415,481 @@ __attribute__((section(".noxip"))) int _imxrt_setLevelLPCG(int clock, int level)
+			*(imxrt_common.anadig_pll + sys_pll3_ctrl) &= ~(1uL << 30u);
+		}
+
+		return;
+	}
+
+	/* Gate all PFDs */
+	*(imxrt_common.anadig_pll + sys_pll3_pfd) |= (1uL << 31u) | (1uL << 23u) | (1uL << 15u) | (1uL << 7u);
+
+	/* Enable, but gate clock */
+	reg = (1u << 4u) | (1uL << 30u);
+	*(imxrt_common.anadig_pll + sys_pll3_pfd) = reg;
+
+	_imxrt_delay(300u * 1000u);
+
+	/* Power off and hold ring off */
+	reg |= (1uL << 21u) | (1uL << 11u);
+	*(imxrt_common.anadig_pll + sys_pll3_pfd) = reg;
+
+	_imxrt_delay(300u * 1000u);
+
+	/* Deassert hold ring off */
+	reg &= ~(1uL << 11u);
+	*(imxrt_common.anadig_pll + sys_pll3_pfd) = reg;
+
+	/* Wait for stable PLL */
+	while ((*(imxrt_common.anadig_pll + sys_pll3_ctrl) & (1uL << 29u)) == 0u) {
+	}
+
+	/* Enable system pll3 and div2 clocks */
+	reg |= (1uL << 13u) | (1uL << 3u);
+
+	/* un-gate sys pll3 */
+	reg &= ~(1uL << 30u);
+	*(imxrt_common.anadig_pll + sys_pll3_pfd) = reg;
+}
+
+
+int _imxrt_setPfdPllFracClock(u8 pfd, u8 clk_pll, u8 frac)
+{
+	volatile u32 *ctrl;
+	volatile u32 *update;
+	volatile u32 stable;
+	u32 fracval;
+	u8 gatedpfd;
+
+	if ((frac < 12u) || (35u < frac)) {
+		return -1;
+	}
+
+	switch (clk_pll) {
+		case clk_pllsys2:
+			ctrl = imxrt_common.anadig_pll + sys_pll2_pfd;
+			update = imxrt_common.anadig_pll + sys_pll2_update;
+			break;
+
+		case clk_pllsys3:
+			ctrl = imxrt_common.anadig_pll + sys_pll3_pfd;
+			update = imxrt_common.anadig_pll + sys_pll3_update;
+			break;
+
+		default:
+			return -1;
+	}
+
+	fracval = ((*ctrl) & (0x3fuL << (8u * pfd))) >> (8u * pfd);
+	gatedpfd = ((*ctrl) & (0x80uL << (8u * pfd))) >> (8u * pfd);
+
+	if ((fracval == (u32)frac) && (gatedpfd == 0u)) {
+		return 0;
+	}
+
+	stable = *ctrl & (0x40uL << (8u * pfd));
+	*ctrl |= 0x80uL << (8u * pfd);
+
+	*ctrl &= ~(0x3fuL << (8u * pfd));
+	*ctrl |= (u32)(frac & 0x3fuL) << (8u * pfd);
+
+	*update ^= 2uL << pfd;
+	*ctrl &= ~(0x80uL << (8u * pfd));
+
+	while ((*ctrl & (0x40uL << (8u * pfd))) == stable) {
+	}
+
+	return 0;
+}
+
+
+void _imxrt_deinitSysPll1(void)
+{
+	/* Disable PLL1 and div2, div5 */
+	*(imxrt_common.anadig_pll + sys_pll1_ctrl) &= ~((1uL << 26u) | (1uL << 25u) | (1uL << 13u));
+
+	/* Gate PLL1 */
+	*(imxrt_common.anadig_pll + sys_pll1_ctrl) |= 1uL << 14u;
+}
+
+
+static void _imxrt_initClockTree(void)
+{
+	static const struct {
+		u8 root;
+		u8 mux;
+		u8 div;
+		u8 isOn;
+	} clktree[] = {
+		{ pctl_clk_m7, mux_clkroot_m7_armpllout, 1, 1 },
+		{ pctl_clk_m4, mux_clkroot_m4_syspll3pfd3, 1, 1 },
+		{ pctl_clk_bus, mux_clkroot_bus_syspll3out, 2, 1 },
+		{ pctl_clk_bus_lpsr, mux_clkroot_bus_lpsr_syspll3out, 3, 1 },
+		/* NOTE: not using "pctl_clk_semc" is disabled by bootrom */
+		{ pctl_clk_cssys, mux_clkroot_cssys_osc24mout, 1, 1 },
+		{ pctl_clk_cstrace, mux_clkroot_cstrace_syspll2out, 4, 1 },
+		{ pctl_clk_m4_systick, mux_clkroot_m4_systick_osc24mout, 1, 1 },
+		{ pctl_clk_m7_systick, mux_clkroot_m7_systick_osc24mout, 2, 1 },
+		{ pctl_clk_adc1, mux_clkroot_adc1_osc24mout, 1, 1 },
```

Comment:
As it seems to make peripheral access very slow (at least in case of GPTx), I wonder why select by default 24 MHz for most of the IP?

Reply:
I used the default settings to be compatible with the default clock settings that are used in the NXP SDK, but I changed the RC-based ones from RCOSC48DIV2 (rc osc) to XTAL-based OSC24M to ensure stability over a wider temperature range (industrial, automotive) and base common clock source.